### PR TITLE
SliverResizingHeader

### PR DESCRIPTION
**Pull request by HansMuller**
A sliver that is pinned to the start of its `CustomScrollView` and reacts to scrolling by resizing between the intrinsic sizes of its min and max extent prototypes.

The minimum and maximum sizes of this sliver are defined by `minExtentPrototype` and `maxExtentPrototype`, a pair of widgets that are laid out once. You can use `SizedBox` widgets to define the size limits.

This sliver is preferable to the general purpose `SliverPersistentHeader` for its relatively narrow use case because there's no need to create a `SliverPersistentHeaderDelegate` or to predict the header's minimum or maximum size.

The sample shows how this sliver's two extent prototype properties can be used to create a resizing header whose minimum and maximum sizes match small and large configurations of the same header widget.

https://github.com/flutter/flutter/assets/1377460/fa7ced98-9d92-4d13-b093-50392118c213

Related sliver utility PRs: https://github.com/flutter/flutter/pull/143538, https://github.com/flutter/flutter/pull/143196, https://github.com/flutter/flutter/pull/127340.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (QuncCccccc) [inline packages/flutter/lib/src/widgets/resizing_header_sliver.dart:61]: Still carefully reviewing all the related PRs🙂. Just FTR, here we probably need to use a `ExcludeFocus` to wrap the `minExtentPrototype` and `maxExtentPrototype`; Otherwise, if we put some tappable widgets here, when we do focus traversal by just using `tab` key, these two would also be traversed, even though they are not really painted. Probably we don't need `IgnorePointer` here because hitTest doesn't include them:)

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/resizing_header_sliver.dart:32]: Can we add a todo to add `See also`s to all these new slivers that can be combined together after they all land?

File: packages/flutter/test/widgets/resizing_header_sliver_test.dart
Comment (Piinks) [inline packages/flutter/test/widgets/resizing_header_sliver_test.dart:10]: Similar comment to the others, there should be a test for each permutation of [vertical, horizontal] : [reversed, not].

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/resizing_header_sliver.dart:61]: Good point, I will fix this stuff and add some tests to check.

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/resizing_header_sliver.dart:52]: Good point. If a max isn't provided it wouldn't be surprising if you got a sliver whose sized matched the min prototype.

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/resizing_header_sliver.dart:33]: If the minimum size defaulted to the child's size (instead of 0) then having this sliver effectively be a PinnedHeaderSliver would work. I'd still prefer to have a PinnedHeaderSliver because it's nice if the class says what you mean, although it could be implemented as a trivial wrapper for this one, or this one's implementation.

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/resizing_header_sliver.dart:43]: Yes.

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/resizing_header_sliver.dart:32]: Yes.

File: packages/flutter/test/widgets/resizing_header_sliver_test.dart
Comment (HansMuller) [reply packages/flutter/test/widgets/resizing_header_sliver_test.dart:10]: Will do.

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/resizing_header_sliver.dart]: I've made the minimum default 0 and the maximum default the intrinsic size of the child. So if you just specify a child you get a header that begins with the child's intrinsic size and shrinks to zero. If the same widget is provided for the minimum and maximum prototypes you get a pinned header.

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/resizing_header_sliver.dart]: The default for maxExtent is now computed differently, see https://github.com/flutter/flutter/pull/143325#discussion_r1626491151

File: packages/flutter/lib/src/widgets/resizing_header_sliver.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/resizing_header_sliver.dart]: I've made the parameter non-null.

**Comment by Piinks** [comment]
Pining into my inbox/review queue :)

File: packages/flutter/lib/src/widgets/sliver_resizing_header.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/sliver_resizing_header.dart:40]: Since this is a pinned sliver, and there is another PR of this set that is pinned and includes `Pinned`, I still feel like it should have `Pinned` in the title, but I'll defer to your judgement.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/sliver_resizing_header.dart:40]: Just to support the idea a bit further, folks searching for Pinned in the API docs will find the other one, and not this option.

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

File: packages/flutter/lib/src/widgets/sliver_resizing_header.dart
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/sliver_resizing_header.dart:40]: Header also means "appears at the top" and I'd really prefer to avoid long compound names. In this case "resizing" emphasizes what's  special about this header, as pinned does for PinnedHeaderSliver. I'll update the see also links in both classes to make sure that if someone finds the wrong header it will be a short trip to find the correct one.

**Pull request merged**
